libqfieldsync/offliners.py

```diff
@@ -0,0 +1,401 @@
+        offline_editing = kwargs.pop("offline_editing", None) or QgsOfflineEditing()
+        super().__init__(*args, **kwargs)
+        self.offliner = offline_editing
+
+        # Check https://api.qgis.org/api/3.14/classQgsOfflineEditing.html#a59d2ebed32704f655868951eba6ef52e for more documentation of these signals
+        # NOTE directly connecting the slot like `self.offliner.progressModeSet.connect(self.progressModeSet)` raises typing error
+        self.offliner.layerProgressUpdated.connect(
+            lambda progress, layer_idx: self.layerProgressUpdated.emit(
+                progress, layer_idx
+            )
+        )
+        self.offliner.progressModeSet.connect(
+            lambda mode, maximum: self.progressModeSet.emit(mode, maximum)
+        )
+        self.offliner.progressUpdated.connect(
+            lambda progress: self.progressUpdated.emit(progress)
+        )
+
+    def convert_to_offline(
+        self,
+        offline_db_filename: str,
+        layers: List[QgsMapLayer],
+        bbox: Optional[QgsRectangle],
+    ) -> bool:
+        offline_db_path = Path(offline_db_filename).parent
+        layer_ids = [layer.id() for layer in layers]
+
+        only_selected = False
+        if bbox and bbox.isFinite():
+            only_selected = True
+            for layer in layers:
+                if Qgis.QGIS_VERSION_INT >= 33000:
+                    no_geometry_types = [
+                        Qgis.GeometryType.Null,
+                        Qgis.GeometryType.Unknown,
+                    ]
+                else:
+                    from qgis.core import QgsWkbTypes
+
+                    no_geometry_types = [
+                        QgsWkbTypes.GeometryType.NullGeometry,
+                        QgsWkbTypes.GeometryType.UnknownGeometry,
+                    ]
+
+                if layer.geometryType() not in no_geometry_types:
+                    # ensure that geometry-less layers do not have selected features that would interfere with the process
+                    layer.removeSelection()
+                else:
+                    layer.selectByRect(bbox)
+
+                if layer.selectedFeatureCount() == 0:
+                    layer.selectByIds([FID_NULL])
+
+        is_success = self.offliner.convertToOfflineProject(
+            str(offline_db_path),
+            str(offline_db_filename),
+            layer_ids,
+            only_selected,
+            # containerType - GPKG or SpatiaLite
+            containerType=QgsOfflineEditing.GPKG,
+            # layerNameSuffix - by default " (offlined)" is added as suffix
+            layerNameSuffix=None,
+        )
+
+        return is_success
+
+
+class PythonMiniOffliner(BaseOffliner):
+    def convert_to_offline(
+        self,
+        offline_db_filename: str,
+        layers: List[QgsMapLayer],
+        bbox: Optional[QgsRectangle],
+    ) -> bool:
+        convert_to_offline_project(str(offline_db_filename), layers, bbox)
```

Comment:
I was in a big doubt if I should go or no for it. My idea was - less OOP, the better, it is slighly more flexible in refactoring, because functions are forced to be as pure as they can be. `OfflineEditing` is OOP and there are gazillion properties on `self` that nobody knows if they are needed. I just deleted 3 of them in this PR...

Reply:
However, not 100% convinced, feel free to insist on doing it full OOP and will do that.

Reply:
I would go for full OOP or otherwise move these pieces into a separate file, the current mix doesn't sit well with me

Reply:
Better have all in one file then. Done!